src/merkle_tree.rs

```diff
@@ -1,27 +1,44 @@
-use std::hash::{DefaultHasher, Hash, Hasher};
+use std::rc::Rc;
+
+use sha3::Digest;
 
 pub enum SiblingHash {
-    Left(u64),
-    Right(u64),
+    Left(Vec<u8>),
+    Right(Vec<u8>),
+}
+
+#[derive(Clone)]
+struct InnerNode {
+    hash_value: Vec<u8>,
+    left_son: Rc<MerkleNode>,
+    right_son: Rc<MerkleNode>,
+}
+#[derive(Clone)]
+struct LeafNode {
+    hash_value: Vec<u8>,
```

Comment:
The leaf node could be parametric and store the data as well:
```rust
#[derive(Clone)]
struct LeafNode<H: AsRef<[u8]> + Clone> {
    hash_value: Vec<u8>,
    date: H,
}
```
Or, if we want to keep things simpler, we could just store the bytes:
```rust
struct LeafNode {
    hash_value: Vec<u8>,
    date: Vec<u8>,
}
```

How you build for each implementation would differ. In the first, you'd `.clone()` while in the second you'd `.as_ref().to_vec()`.

Reply:
But why should we store the data in the `LeafNode` struct? the data is never used, just the `hash_value`

Reply:
It's not strictly necessary, but that way the relationship between leaf nodes and their corresponding data is more obvious.

Reply:
Okay now I see why, should we also change the `leaves` vector from the `MerkleTree` structure to be a `Vec<LeafNode<H>>` instead of just a `Vec<H>`?

Reply:
I would say we could simply remove it if we have the data in the node.

Reply:
You are right, just implemented it without the `leaves` vector